Easy/BestTimetoBuyandSellStock1.java

```diff
@@ -0,0 +1,29 @@
+/**
+ * Best Time to Buy and Sell Stock
+ * Say you have an array for which the ith element is the price of a given stock on day i.
+ * If you were only permitted to complete at most one transaction (ie, buy one and sell one share of the stock), design an algorithm to find the maximum profit.
+ * Tags:  Array Dynamic Programming
+ * Similar Problems:  (M) Maximum Subarray (M) Best Time to Buy and Sell Stock II (H) Best Time to Buy and Sell Stock III (H) Best Time to Buy and Sell Stock IV (M) Best Time to Buy and Sell Stock with Cooldown
+ * O(n) time, O(1) space
+ * @author chenshuna
+ */ 
+
+public class BestTimetoBuyandSellStock {
+    public static int maxProfit(int[] prices) {
+        if(prices == null || prices.length == 0 || prices.length == 1){
+            return 0;
+        }
+        int max = 0;
+        int minPrice = prices[0];
+        for(int i = 0; i < prices.length; i++){
+            minPrice = Math.min(minPrice, prices[i]);
+            max = Math.max(max,  prices[i] - minPrice);
```

Comment:
Just share something fun here. Here `max` is the difference of the maximum and the minimum of the previous prices. Think about this, if `prices[i -1] > prices[i]`, then `maximum of the previous prices > prices[i]`. Then `prices[i] - minPrice` won't be larger than `max`. 

That means, max will only be updated when `prices[i] > prices[i - 1]`.